### ISSUE TITLE
TypedArrays are treated as plain objects

### ISSUE THREAD

**Issue by JobLeonard** [zalmoxisus/redux-devtools-extension#235]
Hey,

Thanks for making this tool, it works really well. However, `TypedArrays` are being displayed as plain objects! Maybe I'm the first to use them together with redux and notice this? It's just a cosmetic thing, so it doesn't matter much, but if you want to fix this: my money is on `Array.isArray` returning `false` for typed arrays; I had to create my own little helper function to deal with that in my own code:

```
export function isArray(obj){
	return obj instanceof Array ||
	obj instanceof Float64Array ||
	obj instanceof Float32Array ||
	obj instanceof Int32Array ||
	obj instanceof Uint32Array ||
	obj instanceof Int16Array ||
	obj instanceof Uint16Array ||
	obj instanceof Int8Array ||
	obj instanceof Uint8Array ||
	obj instanceof Uint8ClampedArray;
}
```

Maybe there's a cleaner way of checking this, but this was the simplest way that worked for me.

**Comment by zalmoxisus** [comment]
That's how `JSON.stringify` interpret typed arrays. So, for example,`JSON.stringify(new Int32Array([101, 102, 103]))` returns `{"0":101,"1":102,"2":103}`.

The extension's API allows you to specify a custom serialization function by using `serializeAction` and `serializeState` like so:
```js
window.devToolsExtension({
  serializeAction: (key, value) => {
    if (value  instanceof Int32Array) {
      var arr = []; 
      for(var p in Object.getOwnPropertyNames(a)) {
        arr[p] = a[p];
      }
      return arr;
   }
    return value;
  }
})
```

Just keep in mind that it will affect the performance significantly.

**Comment by JobLeonard** [comment]
Ah ok, thank you for explaining!

I think I'll stick to being mindful of when "objects" might secretly be typed arrays, since my slow laptop is janky often enough already ;)

**Comment by JobLeonard** [comment]
Hey, if it's OK I'd like to add a follow-up for people who might come across this by googling "redux-devtools-extensions TypedArrays" because their large TypedArrays are crashing the devtools.

I just got around to fixing this issue for myself today, using tricks cobbled together from the documentation and suggestions by @zalmoxisus. See: https://github.com/zalmoxisus/redux-devtools-extension/issues/159#issuecomment-231034408 and https://github.com/zalmoxisus/redux-devtools-extension/blob/master/docs/API/Arguments.md

The trick to avoid this is adding a custom `toJSON` function to the object containing the large array, and make it return an appropriate summary of the array instead.

In the following example, I have wrapped my `data` array in an object containing some useful metadata about it. Both `data` and `uniques` tend to be very large arrays. Furthermore, `data` is usually a typed array. Instead of showing them in full, I display the first five items (transforming `data` to a plain array in the process), and mention the length.I'm using `process.env.NODE_ENV === 'production'` to ensure that toJSON only exists in the debug build:
```JavaScript
	if (process.env.NODE_ENV === 'production') {
		return {
			arrayType,
			data,
			indexedVal,
			uniques,
			colorIndices,
			min,
			max,
		};
	} else {
		// Custom function for redux devtools, to avoid having to serialize
		// the large data arrays (which tends to kill the devtools)
		const toJSON = () => {
			return {
				arrayType,
				data: Array.from(data.slice(0, Math.min(5, data.length))),
				data_length: `${data.length} items`,
				indexedVal,
				uniques: uniques.slice(0, Math.min(5, uniques.length)),
				total_uniques: `${uniques.length} items`,
				colorIndices,
				min,
				max,
			};
		};
		return {
			arrayType,
			data,
			indexedVal,
			uniques,
			colorIndices,
			min,
			max,
			toJSON,
		};
	}
```
